### PR TITLE
Feat: Add string interning on TSDB

### PR DESCRIPTION
**Pull request by vjsamuel**
This PR adds string interning directly on the TSDB. This ensures that samples coming in from both the receiver and the scrape manager benefit from the interning especially on the head. We at eBay, have been running this feature using a `SeriesLifeCycleCallBack` and we have benefited from ~20% memory savings. Cheers to @npordash for the original internal implementation.

This PR adds something similar using a reference counted string pool to intern all strings seen by head. The feature is added behind a feature flag in turned off state.

### CODE REVIEW

**Comment by vjsamuel** [comment]
image

The above image shows ingesting without interning, replaying a WAL without interning and finally replaying the WAL with interning. ~20% difference is seen with and without interning. A total cardinality of 3.4Mil was used for this exercise. 

Doing the opposite:
image

initially ingested 3.5 mil samples with interning turned on and then replayed the WAL with and without interning. ~25% savings seen in using interning. 

The ingest used avalanche. Command used for the same:
`podman run --rm --tls-verify=false  prometheuscommunity/avalanche-linux-amd64:main --remote-url=http://192.168.86.27:9090/api/v1/write --metric-count=1000 --label-count=15 --series-count=100 --series-interval=15 --remote-pprof-urls=http://192.168.86.27:9090/debug/pprof`

**Comment by codesome** [comment]
/prombench main

@bboreham FYI. This PR assumes labels to be a slice that we are moving away from. Still trying out how it works.

File: tsdb/intern.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	}
+}
+
+type interner struct {
+	strings ReferenceCountedStringPool
+}
+
+func (in *interner) Intern(lset labels.Labels) {
+	for i, l := range lset {
+		lset[i] = labels.Label{
```
Comment (codesome) [inline tsdb/intern.go:108]: We are moving away this design of labels and trying to not assume it as a slice anymore. Rather use methods of labels. Reason: https://github.com/prometheus/prometheus/pull/10991.

**Comment by bboreham** [comment]
@vjsamuel please see the design doc linked from #10991 for why I thought this was not worth doing.
If you would like to enhance that design with an interning table that would be great.

File: tsdb/head.go [hunk 2]
```diff
@@ -1749,6 +1764,11 @@ func (s *stripeSeries) getByHash(hash uint64, lset labels.Labels) *memSeries {
 }
 
 func (s *stripeSeries) getOrSet(hash uint64, lset labels.Labels, createSeries func() *memSeries) (*memSeries, bool, error) {
+	// Intern all labels
+	if s.interner != nil {
+		s.interner.Intern(lset)
+	}
```
Comment (colega) [inline tsdb/head.go:1770]: Interning happens before the `PreCreation` hook here, so if that fails we would be leaking those series, right?

**Comment by codesome** [comment]
/prombench cancel

Memory savings are not consistent and not that high (<6% most of the times) compared to what we can get from Bryan's work.

**Comment by roidelapluie** [comment]
We have reviewed this pull request during our bug scrub.

Based on previous comments and the fact that Bryan's work has been merged in prometheus 2.43, we have decided to close this pull request.

Thank you for your contribution.

**Pull request closed without merging**